Rework docs for the `size` of `terms` agg

nik9000 opened the pull request:
The `terms` agg picks the top `size` terms in a single scatter/gather
pass across all the shards. For the default `order` and if you `order`
by `_key` this works quite well. Some errors creep in, but it's fairly
easy to point to them and understand them. But ordering by doc count
ascending is like inviting the error vampire into your agg. It's super
easy to get inaccurate results. This updates the docs to be more stark
about it.

Closes #72684

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 committed:
Rework docs for the `size` of `terms` agg

The `terms` agg picks the top `size` terms in a single scatter/gather
pass across all the shards. For the default `order` and if you `order`
by `_key` this works quite well. Some errors creep in, but it's fairly
easy to point to them and understand them. But ordering by doc count
ascending is like inviting the error vampire into your agg. It's super
easy to get inaccurate results. This updates the docs to be more stark
about it.

Closes #72684

jrodewig reviewed (changes requested):
Thanks for putting this together, @nik9000.

For this first pass, I focused mostly on the `order` warnings. However, I also left some feedback on the other sections. Don't feel obligated to take all the suggestions.

I'd like to take another look once you feel this is ready for it.

nik9000 committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

nik9000 committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

nik9000 committed:
Merge branch 'master' into docs_terms_size_rework

nik9000 committed:
Fixup

nik9000 committed:
Reword a paragraph

nik9000 committed:
Switch SearchExecutionContext to new createParser

In #79814 I deprecated many variants of XContent#createParser in favor
of some new variants that bundle many of the arguments. This replaces
the variants used by the search execution context with the new variants.

jrodewig reviewed (approved):
Thanks for iterating on this @nik9000. Overall, this LGTM.

I mostly left minor style nits that you can disregard if wanted. The only bigger piece of feedback is in the `size` section. We may want to provide clearer guidance on:

- How many terms are "too many" when choosing between the terms agg and the composite agg.
- What happens if you try using a huge `size` on a high-cardinality field.

nik9000 committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <40268737+jrodewig@users.noreply.github.com>

nik9000 committed:
Merge branch 'master' into docs_terms_size_rework

nik9000 committed:
update guidance on "too many"

nik9000 commented:
run elasticsearch-ci/part-2

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.15  | Commit could not be cherrypicked due to conflicts |
| ✅ |  [8.0](https://github.com/elastic/elasticsearch/pull/80162)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/80162">](https://github.com/elastic/elasticsearch/pull/80162) |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 79205`

nik9000 commented:
OK! I'll backport it myself.

nik9000 commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:
```diff
@@ -16,29 +16,29 @@
 import org.elasticsearch.index.mapper.DateFieldMapper;
 import org.elasticsearch.index.shard.IndexLongFieldRange;
-import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
```
Oh yikes this snuck in here! That was not my intent....

nik9000 replied:
It's supposed to be in https://github.com/elastic/elasticsearch/pull/80161 - which was approved. So I'm not going to revert.

nik9000 replied:
But I won't backport it.

nik9000 commented:
Backported in #80223, #80226, and #80227.
